accepted/0000-caasp-nodes.md

```diff
@@ -0,0 +1,326 @@
+* Issue any power management action via Cobbler
+* Perform an SP migration
+
+#### Forbidden packages related actions
+
+Special handling needs to be taken with the packages that are included in the `SUSE-CaaSP-Node` pattern:
+
+- `caasp-config`
+- `cri-o`
+- `kubernetes-client`
+- `kubernetes-kubeadm`
+- `kubernetes-kubelet`
+- `patterns-base-basesystem`
+- `patterns-caasp-Node-1.15`
+- `skuba-update`
+- `supportutils-plugin-suse-caasp`
+
+and their dependencies<sup>3</sup>. At the time of writing, it is not easy to [build the dependency tree using plain `zypper`](https://stackoverflow.com/questions/12183757/reverse-dependency-generation-with-zypper) (it will be available with `zypper` 1.14.33) and it is not clear (from the CaaS Platform side) whether a forbidden package dependency should be considered forbidden as well. For example, `iptables` is a dependency of `cri-o`, but `cri-o` does not specify which versions of `iptables` are supported.
+
+To be on the safe side, Uyuni/SUSE Manager must not offer the possibility to issue the following actions on a registered CaaS Platform node:
+
+* Upgrade a package
+* Remove a package
+
+NOTE: [installing NEW (not already installed and not conflicting with already installed) packages is allowed](https://documentation.suse.com/suse-caasp/4/single-html/caasp-admin/#_existing_cluster).
+
+### Implementation details
+
+Let's now describe how Uyuni/SUSE Manager implements each of the aforementioned actions for a Salt client.
+
+#### System reboot
+
+A reboot is triggered by calling the Salt module `system.reboot` on the target minion.
+
+#### Patch apply
+
+A patch apply is scheduled using Salt `state.apply` with the [patch install state](https://github.com/uyuni-project/uyuni/blob/master/susemanager-utils/susemanager-sls/salt/packages/patchinstall.sls).
+
+#### Mark a system to automatically install patches
+
+Automatic patch installation is implemented at the Uyuni/SUSE Manager level by automatically apply a patch whenever it is available (this action falls into the above one).
+
+#### Issue any power management action via Cobbler
+
+All the power management actions are scheduled via [Cobbler commands](https://cobbler.github.io/manuals/2.8.0/3/1/3_-_Systems.html).
+
+#### Perform an SP migration
+
+An SP migration is scheduled using Salt `state.apply` with the [distupgrade state](https://github.com/uyuni-project/uyuni/blob/master/susemanager-utils/susemanager-sls/salt/distupgrade/init.sls).
+
+#### Package upgrade
+
+A package upgrade is scheduled using Salt `state.apply` with the [pkginstall state](https://github.com/uyuni-project/uyuni/blob/master/susemanager-utils/susemanager-sls/salt/packages/pkginstall.sls).
+
+As a special case when issuing a package upgrade, a package installation from API is scheduled using Salt `state.apply` with the [pkginstall state](https://github.com/uyuni-project/uyuni/blob/master/susemanager-utils/susemanager-sls/salt/packages/pkginstall.sls).
```

Comment:
Do I get correctly that a package upgrade results in a different state being applied depending on whether it is scheduled via API or UI?

Reply:
No. A package upgrade is always issued with a `state.apply packages.pkginstall`.
The difference is that it is not possible to install a non-installed package on the target system from the UI (you can only select a non-installed package from the list).
But it is possible to upgrade an installed package using `system_installpackage` from the API (see https://github.com/uyuni-project/uyuni-rfc/pull/26/files#diff-014d415bc2076bf85e73014398351426R128-R138 )